public/app/features/alerting/unified/components/notification-policies/useNotificationPolicyRoute.ts

```diff
@@ -0,0 +1,162 @@
+        return {
+          ...result,
+          currentData: result.currentData ? k8sRoutesToRoutesMemoized(result.currentData.items) : undefined,
+          data: result.data ? k8sRoutesToRoutes(result.data.items) : undefined,
+        };
+      },
+    }
+  );
+
+  const amConfigQuery = useGetAlertmanagerConfigurationQuery(alertmanager, {
+    skip: skip || k8sApiSupported,
+    selectFromResult: (result) => {
+      return {
+        ...result,
+        currentData: result.currentData?.alertmanager_config?.route
+          ? [result.currentData.alertmanager_config.route]
+          : undefined,
+        data: result.data?.alertmanager_config?.route ? result.data.alertmanager_config.route : undefined,
+      };
+    },
+  });
+
+  return k8sApiSupported ? k8sRouteQuery : amConfigQuery;
+};
+
+export function useUpdateNotificationPolicyRoute(selectedAlertmanager: string) {
+  const [getAlertmanagerConfiguration] = useLazyGetAlertmanagerConfigurationQuery();
+  const [updateAlertmanagerConfiguration] = useUpdateAlertmanagerConfigurationMutation();
+
+  const [updatedNamespacedRoute] = useReplaceNamespacedRoutingTreeMutation();
+
+  const k8sApiSupported = shouldUseK8sApi(selectedAlertmanager);
+
+  async function updateUsingK8sApi({ newRoute }: { newRoute: Route }) {
+    const namespace = getK8sNamespace();
+    const { routes, ...rest } = newRoute;
+    const { provenance, ...defaults } = rest;
+    // Convert Route to K8s compatible format
+    const k8sRoute: ComGithubGrafanaGrafanaPkgApisAlertingNotificationsV0Alpha1RoutingTreeSpec = {
+      defaults: {
+        ...defaults,
+        // TODO: Fix types in k8s API? Fix our types to not allow empty receiver? TBC
+        receiver: defaults.receiver || '',
+      },
+      routes: newRoute.routes?.map(routeToK8sSubRoute) || [],
+    };
+
+    // TODO Add a check to see if not updated in the meantime
+
+    // Create the K8s route object
+    const routeObject = { spec: k8sRoute, metadata: { name: ROOT_ROUTE_NAME } };
+
+    return updatedNamespacedRoute({
+      name: ROOT_ROUTE_NAME,
+      namespace,
+      comGithubGrafanaGrafanaPkgApisAlertingNotificationsV0Alpha1RoutingTree: routeObject,
+    }).unwrap();
+  }
+
+  async function updateUsingConfigFileApi({ newRoute, oldRoute }: { newRoute: Route; oldRoute: Route }) {
+    const lastConfig = await getAlertmanagerConfiguration(selectedAlertmanager).unwrap();
+
+    const configChangedInMeantime = JSON.stringify(oldRoute) !== JSON.stringify(lastConfig.alertmanager_config.route);
+
+    if (configChangedInMeantime) {
+      throw new Error(
+        'A newer Alertmanager configuration is available. Please reload the page and try again to not overwrite recent changes.'
+      );
+    }
+
+    const newConfig = {
+      ...lastConfig,
+      alertmanager_config: {
+        ...lastConfig.alertmanager_config,
+        route: newRoute,
+      },
+    };
+
+    // TODO This needs to properly handle lazy AM initialization
```

Comment:
@konrad147 You left this TODO - how can I test this lazy behaviour/make it happen locally?

Reply:
I _believe_ we tested this last time by running Mimir locally with the `Alertmananger` target enabled – I can help set this up :)